List group keys

paulhenrich opened the pull request:
Add support for grouped listing of keys and object metadata.

The entry point into this functionality is the `riak_client: list_group_keys/4` function, which accepts a `GroupParams` argument to control behavior, such as the maximum number of keys to return, the prefix to which keys are constrained, etc.

The functionality is implemented by the `riak_kv_group_keys_fsm`, which is a `riak_core_coverage_fsm` implementation, and which allows for gathering results from the various nodes using a coverage query. At the vnode level, the grouped key listing is performed by the `riak_kv_group_keys` module, which currently works only for eleveldb.

Commit:
Handle common prefixes

Commit:
Start collating results of list_groupkeys in client

jvoegele committed:
Seek past items not included in result set

When a common prefix is identified, move the iterator to the next key
that is lexicographically after the prefix.

Commit:
Return key as well as metadata in list_group_keys

jvoegele committed:
Handle empty binary in common prefix

jvoegele committed:
Remove backpressure from list_group_keys

jvoegele committed:
Collate results in riak_kv_groupkeys_fsm instead of riak_client

Commit:
Use builtin ordering of ordset instead of sorting in group_keys fsm

Commit:
Add missing function head to continue iteration in prefix-less list_group_keys

Commit:
s/groupkeys/group_keys

Commit:
Seek to correct position after prefix

Commit:
Add max_keys to list_group_keys

Commit:
Avoid returning dupe common prefixes and refactor positional args

into named accumulators in the state.
Dupe CommonPrefixes were caused by inclusion of unique {Key, CommonPrefix} tuple
in the ordset:add_element. Now we return {<<key_or_common_prefix>> [...]}
to achieve uniq|sort.

Commit:
Better interface for group params so we avoid passing around proplist

Commit:
Give back results in correct order and limit by Max Keys

Commit:
Return IsTruncated for list_group_keys

Commit:
Encapsulate group keys response record

Commit:
Refactor to get rid of nested case

Commit:
Basic implementation of list_group_keys for eleveldb backend

Add comment about careful `try...of` usage.

Make explicit that accumulate_contents filters on bucket via pattern matching

Re-indent

jvoegele committed:
Add accessors for #group_params fields

Also, add continuation_token to #group_params and remove some extraneous logging.

jvoegele committed:
Add support for continuation tokens

* Create new module `riak_kv_continuation` for creating and decoding
  continuation tokens.
* Update `riak_kv_group_keys` to use the key that is encoded in the
  continuation token as its starting point.
* In `riak_kv_group_keys_fsm`, generate a continuation token when
  necessary and send it back as part of the reply.

jvoegele committed:
Handle start_after and continuation_token

Add new `start_key` function to uniformly handle both start_after and
continuation_token. Testing revealed that we were incorrectly appending
0x00 instead of 0xFF to the end of the key, which has now been
rectified.

jvoegele committed:
Add setter functions for group_params

jvoegele committed:
Fix Dialyzer errors and warnings

jvoegele committed:
Handle timeout properly in list_group_keys

Also add type spec for `riak_client:list_group_keys/4`.

jvoegele committed:
Filter is not a tagged tuple anymore

This was left over from an early experiment that has since been reverted.

jvoegele committed:
Add basic stats to riak_kv_group_keys_fsm_sup

jvoegele committed:
Add more stats to riak_kv_group_keys_fsm

jvoegele committed:
Address some TODO comments

jvoegele commented on src/riak_kv_group_keys_response.erl:
```diff
@@ -0,0 +1,70 @@
+
+-export_type([response/0]).
+
+-record(response, {
```
Use a more descriptive name for this record to aid debugging.

jvoegele committed:
Use fold_objects instead of fold_keys for grouped fold

nickelization commented on src/riak_kv_continuation.erl:
```diff
@@ -0,0 +1,118 @@
+%% `Results' is greater than `MaxResults'. If so, the item in the list at
+%% position `MaxResults' is used as the term to make the continuation token,
+%% after first being passed to the given `Fun' to allow for transforming the
+%% value. If not, `undefined' is returned.
```
I've been trying to get away from this kind of use of `undefined` lately; I feel like `undefined` is overloaded to the point where its use ends up being kind of ambiguous and error-prone. If we could change it to return `'no_continuation'` or something along those lines, I think that would be much nicer.

nickelization commented on src/riak_kv_continuation.erl:
```diff
@@ -0,0 +1,118 @@
+
+%% @doc Make a continuation token for the given `Term'.
+-spec make_token(term()) -> token().
+make_token(Term) ->
```
I wonder if it might be worth adding an extra byte to the beginning of the token to denote format/version, in case we decide to change to something else later. If nothing else, it'd be nice for efficiency if we could base64-encode raw binaries directly instead of having to wrap them in the extra layer of `term_to_binary`.

On the flip side, this might be overkill though, since I'm not sure if tokens are something we'll be using heavily enough to warrant this level of optimization. I guess use your own judgement :)

nickelization commented on src/riak_kv_continuation.erl:
```diff
@@ -0,0 +1,118 @@
+%% ====================
+%% TESTS
+%% ====================
+
```
Ooh man, perfect opportunity for a little QuickCheck, just to verify that encode/decode round trip :). The tests here look like they're probably sufficient already though, so totally up to you if you want to add in additional testing with EQC.

nickelization commented on src/riak_kv_group_keys.erl:
```diff
@@ -0,0 +1,547 @@
+set_continuation_token(GroupParams, ContinuationToken) ->
+    GroupParams#group_params{continuation_token = ContinuationToken}.
+
+%% NOTE: This  type spec with `fun(() -> no_return())' as the return type is
```
Why is the `no_return()` type necessary? That doesn't quite smell right to me....

Even if there's no other better solution, it'd be nice to have some extra context in the comments as to what's going on here.

nickelization commented on src/riak_kv_group_keys.erl:
```diff
@@ -0,0 +1,547 @@
+              throw(Error)
+    end.
+
+iterator_open(DbRef, FoldOpts) ->
```
Could we maybe get rid of this extra layer of indirection for `iterator_open`, `iterator_close`, and `iterator_move`? Feels kind of like overkill just to avoid having to type "eleveldb:" in a few places.

paulhenrich replied:
Our goal was to start isolating this so it can be used with the memory backend as well, though I agree it looks a little odd as-is.

nickelization commented on src/riak_kv_group_keys.erl:
```diff
@@ -0,0 +1,547 @@
+%% 0xFF is the biggest byte for the purposes of sorting in LevelDB,
+%% used for seeking past all keys starting with the prefix.
+-spec append_max_byte(binary()) -> binary().
+append_max_byte(Binary) when is_binary(Binary) ->
```
But...is it possible someone could write a key starting with `<<Binary/binary, 16#FF>>` into Riak? Then instead of seeking past the prefix, it would seek to the first key starting with the specified prefix followed by 0xff.

Maybe there is something that prevents you from using bytes like this in keys when doing a put. I don't have time to check now, so maybe this is totally safe, but just wanted to raise this possibility.

paulhenrich replied:
I think we decided that in this case it just breaks the optimization a bit of scanning past keys, so in the unlikely case a user writes a key like that we accept it. (it doesn't change the behavior of the consumer, just the efficiency of part of the scan)

Commit:
Should complete renaming of `group_list_keys` to `group_list`.

Really, the group list functionality was in fact iterating over objects,
not just keys, so this is a more appropriate naming convention. Also,
it will allow us to implement the `keys` version of the functionality
later with minimal effort.

paulhenrich commented:
👏🏼
